-Ytest-pickler AssertionError in readLengthType

Test code (based on tests/pos/avoid.scala):
```scala
trait M
object Test2 {
  class NK {
    class I
    type T
  }
  val y = {
    class C extends NK { type T = I }
    new C with M
  }
}
```

`dotc -Ytest-pickler Test.scala`:

```
exception occurred while compiling Test.scala
java.lang.AssertionError: assertion failed: Addr(112) Addr(128) Addr(132) ANDtype while compiling Test.scala
Exception in thread "main" java.lang.AssertionError: assertion failed: Addr(112) Addr(128) Addr(132) ANDtype
	at scala.Predef$.assert(Predef.scala:219)
	at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readLengthType$1(TreeUnpickler.scala:367)
	at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readType(TreeUnpickler.scala:409)
	at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.$anonfun$readType$18(TreeUnpickler.scala:402)
	at scala.collection.mutable.HashMap.getOrElseUpdate(HashMap.scala:82)
	at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readSimpleType$1(TreeUnpickler.scala:402)
	at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readType(TreeUnpickler.scala:409)
	at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readTpt(TreeUnpickler.scala:1206)
	at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readLengthTerm$1(TreeUnpickler.scala:1062)
	at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readTerm(TreeUnpickler.scala:1179)
	at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readLengthTerm$1(TreeUnpickler.scala:1070)
	at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readTerm(TreeUnpickler.scala:1179)
	at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readLengthTerm$1(TreeUnpickler.scala:1070)
	at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readTerm(TreeUnpickler.scala:1179)
	at dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.$anonfun$readNewDef$4(TreeUnpickler.scala:760)
	at dotty.tools.dotc.core.tasty.TreeUnpickler$LazyReader.complete(TreeUnpickler.scala:1290)
	at dotty.tools.dotc.ast.Trees$WithLazyField.forceIfLazy(Trees.scala:869)
	at dotty.tools.dotc.ast.Trees$WithLazyField.forceIfLazy$(Trees.scala:867)
	at dotty.tools.dotc.ast.Trees$ValDef.forceIfLazy(Trees.scala:725)
	at dotty.tools.dotc.ast.Trees$ValOrDefDef.rhs(Trees.scala:376)
	at dotty.tools.dotc.ast.Trees$ValOrDefDef.rhs$(Trees.scala:376)
	at dotty.tools.dotc.ast.Trees$ValDef.rhs(Trees.scala:725)
	at dotty.tools.dotc.printing.RefinedPrinter.$anonfun$valDefToText$2(RefinedPrinter.scala:679)
	at dotty.tools.dotc.printing.RefinedPrinter.withEnclosingDef(RefinedPrinter.scala:45)
	at dotty.tools.dotc.printing.RefinedPrinter.$anonfun$valDefToText$1(RefinedPrinter.scala:679)
	at dotty.tools.dotc.printing.RefinedPrinter.dclTextOr(RefinedPrinter.scala:665)
	at dotty.tools.dotc.printing.RefinedPrinter.valDefToText(RefinedPrinter.scala:678)
	at dotty.tools.dotc.printing.RefinedPrinter.toTextCore(RefinedPrinter.scala:451)
	at dotty.tools.dotc.printing.RefinedPrinter.$anonfun$toText$18(RefinedPrinter.scala:571)
	at dotty.tools.dotc.printing.PlainPrinter.controlled(PlainPrinter.scala:27)
	at dotty.tools.dotc.printing.RefinedPrinter.toText(RefinedPrinter.scala:568)
	at dotty.tools.dotc.ast.Trees$Tree.toText(Trees.scala:211)
	at dotty.tools.dotc.printing.Printer.$anonfun$toText$1(Printer.scala:153)
	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:233)
	at scala.collection.immutable.List.foreach(List.scala:388)
	at scala.collection.TraversableLike.map(TraversableLike.scala:233)
	at scala.collection.TraversableLike.map$(TraversableLike.scala:226)
	at scala.collection.immutable.List.map(List.scala:294)
	at dotty.tools.dotc.printing.Printer.toText(Printer.scala:153)
	at dotty.tools.dotc.printing.Printer.$anonfun$toTextGlobal$2(Printer.scala:161)
	at dotty.tools.dotc.printing.Printer.atPrec(Printer.scala:41)
	at dotty.tools.dotc.printing.Printer.toTextGlobal(Printer.scala:161)
	at dotty.tools.dotc.printing.RefinedPrinter.toTextTemplate(RefinedPrinter.scala:722)
	at dotty.tools.dotc.printing.RefinedPrinter.$anonfun$templateText$1(RefinedPrinter.scala:729)
	at dotty.tools.dotc.printing.RefinedPrinter.withEnclosingDef(RefinedPrinter.scala:45)
	at dotty.tools.dotc.printing.RefinedPrinter.templateText(RefinedPrinter.scala:729)
	at dotty.tools.dotc.printing.RefinedPrinter.recur$1(RefinedPrinter.scala:463)
	at dotty.tools.dotc.printing.RefinedPrinter.toTextCore(RefinedPrinter.scala:473)
	at dotty.tools.dotc.printing.RefinedPrinter.$anonfun$toText$18(RefinedPrinter.scala:571)
	at dotty.tools.dotc.printing.PlainPrinter.controlled(PlainPrinter.scala:27)
	at dotty.tools.dotc.printing.RefinedPrinter.toText(RefinedPrinter.scala:568)
	at dotty.tools.dotc.ast.Trees$Tree.toText(Trees.scala:211)
	at dotty.tools.dotc.printing.Printer.$anonfun$toText$1(Printer.scala:153)
	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:233)
	at scala.collection.immutable.List.foreach(List.scala:388)
	at scala.collection.TraversableLike.map(TraversableLike.scala:233)
	at scala.collection.TraversableLike.map$(TraversableLike.scala:226)
	at scala.collection.immutable.List.map(List.scala:294)
	at dotty.tools.dotc.printing.Printer.toText(Printer.scala:153)
	at dotty.tools.dotc.printing.Printer.$anonfun$toTextGlobal$2(Printer.scala:161)
	at dotty.tools.dotc.printing.Printer.atPrec(Printer.scala:41)
	at dotty.tools.dotc.printing.Printer.toTextGlobal(Printer.scala:161)
	at dotty.tools.dotc.printing.RefinedPrinter.packageDefText(RefinedPrinter.scala:740)
	at dotty.tools.dotc.printing.RefinedPrinter.toTextCore(RefinedPrinter.scala:485)
	at dotty.tools.dotc.printing.RefinedPrinter.$anonfun$toText$18(RefinedPrinter.scala:571)
	at dotty.tools.dotc.printing.PlainPrinter.controlled(PlainPrinter.scala:27)
	at dotty.tools.dotc.printing.RefinedPrinter.toText(RefinedPrinter.scala:568)
	at dotty.tools.dotc.ast.Trees$Tree.toText(Trees.scala:211)
	at dotty.tools.dotc.printing.Showable.show(Showable.scala:23)
	at dotty.tools.dotc.printing.Showable.show$(Showable.scala:23)
	at dotty.tools.dotc.ast.Trees$Tree.show(Trees.scala:54)
	at dotty.tools.dotc.printing.Formatting$StringFormatter.showArg(Formatting.scala:29)
	at dotty.tools.dotc.printing.Formatting$StringFormatter.$anonfun$treatArg$2(Formatting.scala:43)
	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:233)
	at scala.collection.immutable.List.foreach(List.scala:388)
	at scala.collection.TraversableLike.map(TraversableLike.scala:233)
	at scala.collection.TraversableLike.map$(TraversableLike.scala:226)
	at scala.collection.immutable.List.map(List.scala:294)
	at dotty.tools.dotc.printing.Formatting$StringFormatter.treatArg(Formatting.scala:43)
	at dotty.tools.dotc.printing.Formatting$StringFormatter.$anonfun$assemble$3(Formatting.scala:59)
	at scala.runtime.Tuple2Zipped$.$anonfun$map$1(Tuple2Zipped.scala:47)
	at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
	at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
	at scala.collection.mutable.WrappedArray.foreach(WrappedArray.scala:37)
	at scala.runtime.Tuple2Zipped$.map$extension(Tuple2Zipped.scala:45)
	at dotty.tools.dotc.printing.Formatting$StringFormatter.assemble(Formatting.scala:59)
	at dotty.tools.dotc.core.Decorators$StringInterpolators$.i$extension(Decorators.scala:197)
	at dotty.tools.dotc.transform.Pickler.$anonfun$testUnpickler$6(Pickler.scala:107)
	at dotty.tools.dotc.transform.Pickler.$anonfun$testUnpickler$6$adapted(Pickler.scala:105)
	at scala.collection.TraversableLike$WithFilter.$anonfun$foreach$1(TraversableLike.scala:788)
	at scala.collection.mutable.HashMap.$anonfun$foreach$1(HashMap.scala:145)
	at scala.collection.mutable.HashTable.foreachEntry(HashTable.scala:235)
	at scala.collection.mutable.HashTable.foreachEntry$(HashTable.scala:228)
	at scala.collection.mutable.HashMap.foreachEntry(HashMap.scala:40)
	at scala.collection.mutable.HashMap.foreach(HashMap.scala:145)
	at scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:787)
	at dotty.tools.dotc.transform.Pickler.testUnpickler(Pickler.scala:105)
	at dotty.tools.dotc.transform.Pickler.runOn(Pickler.scala:91)
	at dotty.tools.dotc.Run.$anonfun$compileUnits$3(Run.scala:172)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at dotty.tools.dotc.util.Stats$.trackTime(Stats.scala:49)
	at dotty.tools.dotc.Run.$anonfun$compileUnits$2(Run.scala:169)
	at dotty.tools.dotc.Run.$anonfun$compileUnits$2$adapted(Run.scala:167)
	at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
	at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:194)
	at dotty.tools.dotc.Run.runPhases$1(Run.scala:167)
	at dotty.tools.dotc.Run.$anonfun$compileUnits$1(Run.scala:192)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:90)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:147)
	at dotty.tools.dotc.Run.compileSources(Run.scala:134)
	at dotty.tools.dotc.Run.compile(Run.scala:118)
	at dotty.tools.dotc.Driver.doCompile(Driver.scala:30)
	at dotty.tools.dotc.Driver.process(Driver.scala:136)
	at dotty.tools.dotc.Driver.process(Driver.scala:105)
	at dotty.tools.dotc.Driver.process(Driver.scala:117)
	at dotty.tools.dotc.Driver.main(Driver.scala:144)
	at dotty.tools.dotc.Main.main(Main.scala)
```

--- a/compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala
+++ b/compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala
@@ -392,7 +392,13 @@ class TreeUnpickler(reader: TastyReader,
         case THIS =>
           ThisType.raw(readType().asInstanceOf[TypeRef])
         case RECtype =>
-          typeAtAddr.getOrElse(start, RecType(rt => registeringType(rt, readType())))
+          typeAtAddr.get(start) match {
+            case Some(tp) =>
+              skipTree(tag)
+              tp
+            case None =>
+              RecType(rt => registeringType(rt, readType()))
+          }
         case RECthis =>
           readTypeRef().asInstanceOf[RecType].recThis
         case TYPEALIAS =>

--- a/compiler/test/dotc/pos-decompilation.blacklist
+++ b/compiler/test/dotc/pos-decompilation.blacklist
@@ -12,3 +12,6 @@ tcpoly_checkkinds_mix.scala
 i3050.scala
 i4006b.scala
 i4006c.scala
+
+# Decompiling RecThis as "this" is incorrect
+avoid.scala

--- a/tests/pos/avoid.scala
+++ b/tests/pos/avoid.scala
@@ -9,6 +9,8 @@ object test {
   val z: String = x.y
 }
 
+trait M
+
 // A tricky case involving inner classes, exercised
 // in the large in dotty.tools.dotc.core.NameKinds.scala.
 object Test2 {
@@ -23,5 +25,10 @@ object Test2 {
   val y = {
     class C extends NK { type T = I }
     new C
+  }
+
+  val z = {
+    class C extends NK { type T = I }
+    new C with M
   }
 }

